# TV shows Popularity Predictor (39%)

The goal of this challenge is to create a model that predicts the `popularity` of a movie or TV show

<img src="image.jpg" width=300 />




The dataset contains a list of movies and TV shows with the following characteristics:
- `title`: title of the movie in english
- `original_title`: original title of the movie 
- `duration_min`: duration of the movie in minutes
- `popularity`: popularity of the movie in terms of review scores
- `release_date`: release date
- `description`: short summary of the movie
- `budget`: budget spent to produce the movie in USD
- `revenue`: movie revenue in USD 
- `original_language`: original language 
- `status`: is the movie already released or not
- `number_of_awards_won`: number of awards won for the movie
- `number_of_nominations`: number of nominations
- `has_collection`: if the movie is part of a sequel or not
- `all_genres`: genres that described the movie (can be zero, one or many!) 
- `top_countries`: countries where the movie was produced (can be zero, one or many!) 
- `number_of_top_productions`: number of top production companies that produced the film if any. 
Top production companies includes: Warner Bros, Universal Pictures, Paramount Pictures, Canal+, etc...
- `available_in_english`: whether the movie is available in english or not

## Imports

Run the following cell to load the basic packages:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from nbresult import ChallengeResult

## Data collection

📝 **Load the `movie_popularity.csv` dataset from the provided this [URL](https://wagon-public-datasets.s3.amazonaws.com/certification_france_2021_q2/tv_movies_popularity.csv)**
- First, check and remove the rows that may be complete duplicate from one another (we never know!)
- Then, drop the columns that have too much missing values
- Finally, drop the few remaining rows that have missing values
- Store the result in a `DataFrame` named `data`

In [2]:
# YOUR CODE HERE
import io
import requests
url="https://wagon-public-datasets.s3.amazonaws.com/certification_france_2021_q2/tv_movies_popularity.csv"
s=requests.get(url).content
data=pd.read_csv(io.StringIO(s.decode('utf-8')))

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6864 entries, 0 to 6863
Data columns (total 16 columns):
original_title               6864 non-null object
title                        6863 non-null object
release_date                 6864 non-null object
duration_min                 6864 non-null float64
description                  6864 non-null object
budget                       6864 non-null int64
revenue                      2778 non-null float64
original_language            6864 non-null object
status                       6864 non-null object
number_of_awards_won         6864 non-null int64
number_of_nominations        6864 non-null int64
has_collection               6864 non-null int64
all_genres                   6864 non-null object
top_countries                6864 non-null object
number_of_top_productions    6864 non-null int64
available_in_english         6864 non-null bool
dtypes: bool(1), float64(2), int64(5), object(8)
memory usage: 811.2+ KB


### 🧪 Run the following cell to save your results

In [4]:
from nbresult import ChallengeResult

result = ChallengeResult(
    "data_cleaning",
    columns=data.columns,
    cleaning=sum(data.isnull().sum()),
    shape=data.shape)
result.write()

## Baseline model

### The metric

📝 **We want to predict `popularity`: Start by plotting a histogram of the target to visualize it**

📝 **Which sklearn's scoring [metric](https://scikit-learn.org/stable/modules/model_evaluation.html) should we use if we want it to:**

- Be better when greater (i.e. metric_good_model > metric_bad_model)
- Penalize **more** an error between 10 and 20 compared with an error between 110 and 120
- Said otherwise, what matter should be the **relative error ratio**, more than the absolute error difference

Hint: the histogram plotted above should give you some intuition about it

👉 Store its exact [sklearn scoring name](https://scikit-learn.org/stable/modules/model_evaluation.html) as `string` in the variable `scoring` below.

🚨 You must use this metric for the rest of the challenge

In [93]:
scoring = 'r2'

In [92]:
# YOUR CODE HERE
#plt.hist(data['popularity'],bins = 500)
#plt.show()
sns.histplot(data=data, x="popularity")

ValueError: Could not interpret value `popularity` for parameter `x`

<details>
    <summary>💡 Hint</summary>
It is around here!
<img src="scores.jpg" width=200 height=400 />
</details>

### X,y

**📝 Define `X` as the features Dataframe (keep all features) and `y` as the target Series.**

In [18]:
# YOUR CODE HERE
y= data['popularity']
data = data.drop(['popularity'], axis=1)
X = data
X

original_title  \
0                         Hot Tub Time Machine 2   
1       The Princess Diaries 2: Royal Engagement   
2                                       Whiplash   
3                                        Kahaani   
4                                 The Possession   
5                         Muppet Treasure Island   
6                                  A Mighty Wind   
7                                          Rocky   
8     Revenge of the Nerds II: Nerds in Paradise   
9                                American Beauty   
10                                       Be Cool   
11                               Minority Report   
12                                       Šišanje   
13                                       Red Eye   
14                           The Invisible Woman   
15                                   Chalet Girl   
16                                 Transporter 2   
17                                 Lost in Space   
18                      Friday the 13th Part III   
19                                         V/H/S   
20                                   Black Sheep   
21                                   Dr. Giggles   
22                          The Spanish Prisoner   
23                                       What If   
24                   The Transformers: The Movie   
25                                Changing Lanes   
26                                  Intouchables   
27                                         Caché   
28                         The Last Witch Hunter   
29                               ABCs of Death 2   
...                                          ...   
6834                                          공조   
6835                                Rock of Ages   
6836                                   Planet 51   
6837                                      L'Ours   
6838                                 Constantine   
6839                                    Dogville   
6840                        The Osterman Weekend   
6841                                The Darkness   
6842                             Two Night Stand   
6843                                      Jaws 2   
6844                                    The Debt   
6845                              Romeo & Juliet   
6846                                     Air Bud   
6847                                 The Pianist   
6848                                    Stitches   
6849                               The Informers   
6850                                A Dark Truth   
6851                                     Taken 2   
6852                              Over the Hedge   
6853                          Music of the Heart   
6854                             L'illusionniste   
6855                                     Snowden   
6856                                       Focus   
6857        Friday the 13th Part VI: Jason Lives   
6858                          That Thing You Do!   
6859                              The King and I   
6860                           Little Black Book   
6861                         The English Teacher   
6862                       2001: A Space Odyssey   
6863                               Event Horizon   

                                           title release_date  duration_min  \
0                         Hot Tub Time Machine 2   2015-02-20          93.0   
1       The Princess Diaries 2: Royal Engagement   2004-08-06         113.0   
2                                       Whiplash   2014-10-10         105.0   
3                                        Kahaani   2012-03-09         122.0   
4                                 The Possession   2012-08-30          92.0   
5                         Muppet Treasure Island   1996-02-16         100.0   
6                                  A Mighty Wind   2003-04-16          91.0   
7                                          Rocky   1976-11-21         119.0   
8     Revenge of the Nerds II: Nerds in Paradise   1987-07-10          98.0   
9                  

### Basic pipeline

📝 **Check unique values per features**

In [20]:
# YOUR CODE HERE
for i in X.columns:
    print(X[i].unique())

['Hot Tub Time Machine 2' 'The Princess Diaries 2: Royal Engagement'
 'Whiplash' ... 'The Verdict' 'It Follows'
 'Vivre sa vie: film en douze tableaux']
['Hot Tub Time Machine 2' 'The Princess Diaries 2: Royal Engagement'
 'Whiplash' ... 'The Verdict' 'It Follows' 'Vivre Sa Vie']
['2015-02-20' '2004-08-06' '2014-10-10' ... '2006-01-30' '2015-06-23'
 '1962-09-20']
[ 93. 113. 105. 122.  92. 100.  91. 119.  98. 118. 145.  97.  85. 111.
  96.  87. 130.  95. 116. 110.  84.  89. 112. 117. 106. 125.  94. 127.
 123. 126. 108.  88. 167. 102. 160. 107. 144. 124. 115. 133. 129. 104.
 103. 157. 109. 135. 147. 120. 121. 177. 189. 178.  86.  90.  79. 101.
  72.  99. 141. 136. 143. 132. 139. 114. 140.  83. 148. 137. 168. 156.
 154.  77.  76. 163. 155. 150. 149. 170.  80. 128. 131. 220. 181. 179.
  81.  82. 161. 134. 193. 158. 188. 212. 142. 146. 151. 171. 165. 162.
 185. 138.  78.  75. 153. 175. 219. 183. 186. 152. 199.  68. 214. 248.
 159. 180. 187.  73. 197. 164. 169. 172.  66.  63. 213. 174.  69. 

In this baseline, let's forget about the columns below that are difficult to process

In [50]:
text = ['description', 'original_title', 'title']
dates = ['release_date'] 

We will simply scale the numerical features and one-hot-encode the categorical ones remaining

📝 **Prepare 2 `list`s of features names as `str`**:
- `numerical` which contains **only** numerical features
- `categorical` which contains **only** categorical features (exept text and dates above)

In [23]:
# YOUR CODE HERE
numerical = ['duration_min','budget','revenue','number_of_awards_won', 'number_of_nominations','number_of_top_productions']
categorical = ['original_language','status','has_collection','all_genres', 'top_countries','available_in_english']
X.columns

Index(['original_title', 'title', 'release_date', 'duration_min',
       'description', 'budget', 'revenue', 'original_language', 'status',
       'number_of_awards_won', 'number_of_nominations', 'has_collection',
       'all_genres', 'top_countries', 'number_of_top_productions',
       'available_in_english'],
      dtype='object')

In [51]:
X = X.drop(text, axis=1)
X = X.drop(dates, axis=1)
X

duration_min     budget      revenue original_language    status  \
0             93.0   14000000   12314651.0                en  Released   
1            113.0   40000000   95149435.0                en  Released   
2            105.0    3300000   13092000.0                en  Released   
3            122.0    1200000   16000000.0                hi  Released   
4             92.0   14000000   85446075.0                en  Released   
5            100.0          0   34327391.0                en  Released   
6             91.0    6000000   18750246.0                en  Released   
7            119.0    1000000  117235147.0                en  Released   
8             98.0          0   22642033.0                en  Released   
9            122.0   15000000  356296601.0                en  Released   
10           118.0   53000000   95226116.0                en  Released   
11           145.0  102000000  358372926.0                en  Released   
12            97.0     500000     192864.0                sr  Released   
13            85.0   26000000   57891803.0                en  Released   
14           111.0          0    1234254.0                en  Released   
15            96.0    8000000    1749457.0                en  Released   
16            87.0   32000000   85167639.0                en  Released   
17           130.0   80000000  136159423.0                en  Released   
18            95.0    4000000   36690067.0                en  Released   
19           116.0          0     100345.0                en  Released   
20            87.0          0    4947717.0                en  Released   
21            95.0          0    8403433.0                en  Released   
22           110.0   10000000   13835130.0                en  Released   
23            93.0   11000000    7847000.0                en  Released   
24            84.0    6000000    5849647.0                en  Released   
25            89.0   45000000   94935764.0                en  Released   
26           112.0   13000000  426480871.0                fr  Released   
27           117.0          0   36000000.0                fr  Released   
28           106.0   90000000  146936910.0                en  Released   
29           125.0          0       7171.0                en  Released   
...            ...        ...          ...               ...       ...   
6834         125.0    8520000          NaN                ko  Released   
6835         123.0   75000000          NaN                en  Released   
6836          91.0   70000000          NaN                en  Released   
6837          94.0          0          NaN                en  Released   
6838         121.0  100000000  230884728.0                en  Released   
6839         178.0   10000000          NaN                da  Released   
6840         103.0          0          NaN                en  Released   
6841          92.0    4000000          NaN                en  Released   
6842          86.0          0          NaN                en  Released   
6843         116.0   20000000          NaN                en  Released   
6844         113.0   20000000          NaN                en  Released   
6845         118.0          0          NaN                en  Released   
6846          98.0    3000000          NaN                en  Released   
6847         150.0   35000000          NaN                en  Released   
6848          86.0     100000          NaN                en  Released   
6849          98.0   18000000     382174.0                en  Released   
6850         106.0          0          NaN                en  Released   
6851          91.0   45000000          NaN                en  Released   
6852          83.0   80000000          NaN                en  Released   
6853         124.0   27000000   14859394.0                en  Released   
6854          80.0   18000000    5602431.0                fr  Released   
6855         134.0   40000000   37279439.0                en  Released   
6856      

### Pipelining

You are going to build a basic pipeline made of a basic preprocessing and a trees-based model of your choice.

#### Preprocessing pipeline

**📝 Create a basic preprocessing pipeline for the 2 types of features above:**
- It should scale the `numerical` features
- one-hot-encode the `categorical` and `boolean` features
- drop the others
- Store your pipeline in a `basic_preprocessing` variable

In [24]:
# Execute this cell to enable a nice display for your pipelines
from sklearn import set_config; set_config(display='diagram')

In [77]:
# YOUR CODE HERE
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.compose import make_column_selector

# Impute then Scale for numerical variables: 
num_transformer = Pipeline([
    ('imputer', SimpleImputer()),
    ('scaler', StandardScaler())])


# Encode categorical variables
cat_transformer = OneHotEncoder(handle_unknown='ignore',sparse=False)



# Paralellize "num_transformer" and "One hot encoder"
preprocessor = ColumnTransformer([
    ('num_transformer', num_transformer, numerical),
    ('cat_transformer', cat_transformer, categorical)]
)
    



preprocessor

ColumnTransformer(transformers=[('num_transformer',
                                 Pipeline(steps=[('imputer', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['duration_min', 'budget', 'revenue',
                                  'number_of_awards_won',
                                  'number_of_nominations',
                                  'number_of_top_productions']),
                                ('cat_transformer',
                                 OneHotEncoder(handle_unknown='ignore',
                                               sparse=False),
                                 ['original_language', 'status',
                                  'has_collection', 'all_genres',
                                  'top_countries', 'available_in_english'])])

In [95]:
final_pipe = Pipeline([
    ('preprocessing', preprocessor)])

basic_preprocessing = final_pipe

In [76]:

pd.DataFrame(final_pipe.fit_transform(X))

0         1             2         3         4         5     6     \
0    -0.796276 -0.302503 -6.887537e-01 -0.421280 -0.425453  3.547566   0.0   
1     0.182492  0.376600  2.191277e-01 -0.358858 -0.425453  0.532212   0.0   
2    -0.209015 -0.581980 -6.802338e-01  5.633718  5.242737 -0.975465   0.0   
3     0.622938 -0.636831 -6.483617e-01  1.014441  0.208750 -0.975465   0.0   
4    -0.845214 -0.302503  1.127775e-01 -0.421280 -0.266902 -0.975465   0.0   
5    -0.453707 -0.668174 -4.474908e-01 -0.421280 -0.306540  0.532212   0.0   
6    -0.894153 -0.511458 -6.182186e-01  0.452637  0.605127 -0.975465   0.0   
7     0.476123 -0.642055  4.611904e-01  0.827173  0.327664  0.532212   0.0   
8    -0.551584 -0.668174 -5.755641e-01 -0.421280 -0.504728  0.532212   0.0   
9     0.622938 -0.276384  3.081339e+00  6.507635  3.538317  0.532212   0.0   
10    0.427184  0.716151  2.199682e-01 -0.358858  0.010562  0.532212   0.0   
11    1.748522  1.995999  3.104096e+00  0.827173  3.102302  3.547566   0.0   
12   -0.600522 -0.655114 -8.216103e-01 -0.358858 -0.504728 -0.975465   0.0   
13   -1.187783  0.010929 -1.892214e-01 -0.296435 -0.147989  0.532212   0.0   
14    0.084616 -0.668174 -8.101965e-01 -0.358858 -0.227264  0.532212   0.0   
15   -0.649461 -0.459219 -8.045498e-01 -0.421280 -0.465090 -0.975465   0.0   
16   -1.089907  0.167645  1.097258e-01 -0.421280 -0.504728  2.039889   0.0   
17    1.014446  1.421373  6.686032e-01 -0.234012  0.169113  0.532212   0.0   
18   -0.698399 -0.563697 -4.215956e-01 -0.421280 -0.425453  0.532212   0.0   
19    0.329308 -0.668174 -8.226243e-01 -0.421280 -0.306540 -0.975465   0.0   
20   -1.089907 -0.668174 -7.694964e-01 -0.109167 -0.306540 -0.975465   0.0   
21   -0.698399 -0.668174 -7.316212e-01 -0.358858 -0.425453  0.532212   0.0   
22    0.035677 -0.406981 -6.720890e-01 -0.421280 -0.385815 -0.975465   0.0   
23   -0.796276 -0.380861 -7.377198e-01 -0.296435 -0.187626 -0.975465   0.0   
24   -1.236722 -0.511458 -7.596111e-01 -0.421280 -0.425453 -0.975465   0.0   
25   -0.992030  0.507196  2.167859e-01 -0.421280 -0.227264  0.532212   0.0   
26    0.133554 -0.328623  3.850569e+00  1.950781  1.080780  0.532212   0.0   
27    0.378246 -0.668174 -4.291588e-01  1.388977  0.961867  0.532212   0.0   
28   -0.160077  1.682567  7.867261e-01 -0.421280 -0.465090  0.532212   0.0   
29    0.769753 -0.668174 -8.236455e-01 -0.421280 -0.465090 -0.975465   0.0   
...        ...       ...           ...       ...       ...       ...   ...   
6834  0.769753 -0.445637  1.633189e-16 -0.046744 -0.266902 -0.975465   0.0   
6835  0.671877  1.290777  1.633189e-16 -0.296435 -0.187626  0.532212   0.0   
6836 -0.894153  1.160180  1.633189e-16 -0.171590 -0.346177  0.532212   0.0   
6837 -0.747338 -0.668174  1.633189e-16 -0.109167 -0.068713 -0.975465   0.0   
6838  0.574000  1.943760  1.706807e+00 -0.296435 -0.108351  2.039889   0.0   
6839  3.363490 -0.406981  1.633189e-16  0.889596  0.803316 -0.975465   0.0   
6840 -0.306892 -0.668174  1.633189e-16 -0.296435 -0.504728 -0.975465   0.0   
6841 -0.845214 -0.563697  1.633189e-16 -0.421280 -0.504728  0.532212   0.0   
6842 -1.138845 -0.668174  1.633189e-16 -0.421280 -0.465090 -0.975465   0.0   
6843  0.329308 -0.145787  1.633189e-16 -0.421280 -0.504728  0.532212   0.0   
6844  0.182492 -0.145787  1.633189e-16  0.390214 -0.306540  0.532212   0.0   
6845  0.427184 -0.668174  1.633189e-16 -0.296435 -0.425453 -0.975465   0.0   
6846 -0.551584 -0.589816  1.633189e-16 -0.171590 -0.385815  0.532212   0.0   
6847  1.993214  0.246003  1.633189e-16  3.136811  2.428461  2.039889   0.0   
6848 -1.138845 -0.665562  1.633189e-16 -0.296435 -0.504728 -0.975465   0.0   
6849 -0.551584 -0.198026 -8.195354e-01 -0.421280 -0.504728 -0.975465   0.0   
6850 -0.160077 -0.668174  1.633189e-16 -0.358858 -0.504728 -0.975465   0.0   
6851 -0.894153  0.507196  1.633189e-16 -0.358858 -0.385815  2.039889   0.0   
6852 -1.285660  1.421373  1.633189e-16  0.015678  0.525852 -0.975465   0.0   
6853  0.720815  0.037048 -6.6086

In [28]:
X

original_title  \
0                         Hot Tub Time Machine 2   
1       The Princess Diaries 2: Royal Engagement   
2                                       Whiplash   
3                                        Kahaani   
4                                 The Possession   
5                         Muppet Treasure Island   
6                                  A Mighty Wind   
7                                          Rocky   
8     Revenge of the Nerds II: Nerds in Paradise   
9                                American Beauty   
10                                       Be Cool   
11                               Minority Report   
12                                       Šišanje   
13                                       Red Eye   
14                           The Invisible Woman   
15                                   Chalet Girl   
16                                 Transporter 2   
17                                 Lost in Space   
18                      Friday the 13th Part III   
19                                         V/H/S   
20                                   Black Sheep   
21                                   Dr. Giggles   
22                          The Spanish Prisoner   
23                                       What If   
24                   The Transformers: The Movie   
25                                Changing Lanes   
26                                  Intouchables   
27                                         Caché   
28                         The Last Witch Hunter   
29                               ABCs of Death 2   
...                                          ...   
6834                                          공조   
6835                                Rock of Ages   
6836                                   Planet 51   
6837                                      L'Ours   
6838                                 Constantine   
6839                                    Dogville   
6840                        The Osterman Weekend   
6841                                The Darkness   
6842                             Two Night Stand   
6843                                      Jaws 2   
6844                                    The Debt   
6845                              Romeo & Juliet   
6846                                     Air Bud   
6847                                 The Pianist   
6848                                    Stitches   
6849                               The Informers   
6850                                A Dark Truth   
6851                                     Taken 2   
6852                              Over the Hedge   
6853                          Music of the Heart   
6854                             L'illusionniste   
6855                                     Snowden   
6856                                       Focus   
6857        Friday the 13th Part VI: Jason Lives   
6858                          That Thing You Do!   
6859                              The King and I   
6860                           Little Black Book   
6861                         The English Teacher   
6862                       2001: A Space Odyssey   
6863                               Event Horizon   

                                           title release_date  duration_min  \
0                         Hot Tub Time Machine 2   2015-02-20          93.0   
1       The Princess Diaries 2: Royal Engagement   2004-08-06         113.0   
2                                       Whiplash   2014-10-10         105.0   
3                                        Kahaani   2012-03-09         122.0   
4                                 The Possession   2012-08-30          92.0   
5                         Muppet Treasure Island   1996-02-16         100.0   
6                                  A Mighty Wind   2003-04-16          91.0   
7                                          Rocky   1976-11-21         119.0   
8     Revenge of the Nerds II: Nerds in Paradise   1987-07-10          98.0   
9                  

**📝 Encode the features and store the result in the variable `X_basic_preprocessing`.**

In [78]:
# YOUR CODE HERE
X_basic_preprocessing = pd.DataFrame(final_pipe.fit_transform(X))

**❓ How many features has been generated by the preprocessing? What do you think about this number?**

In [84]:
X_basic_preprocessing.shape

(6864, 1120)

> YOUR ANSWER HERE

 1120 feature have been generated cause of the one hot encoding of feature with a lot a unique value. This number is a little high

#### Modeling pipeline

Let's add a model to our pipe. With so many features one-hot-encoded, we **need a model which can act as a feature selector**

👉 A linear model regularized with L1 penalty is a good starting point.


**📝 Create a `basic_pipeline` which encapsulate the `basic_preprocessing` pipeline + a linear model with a L1 penalty**

- store the resulting pipeline as `basic_pipeline`
- don't fine-tune it


<details>
    <summary>Hints</summary>

Choose your model from the list [here](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.linear_model)

</details>

In [97]:
# YOUR CODE HERE

from sklearn.linear_model import Ridge

final_pipe = Pipeline([
    ('preprocessing', preprocessor),
    ('linear_regression', Ridge())])
basic_pipeline = final_pipe
final_pipe

Pipeline(steps=[('preprocessing',
                 ColumnTransformer(transformers=[('num_transformer',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['duration_min', 'budget',
                                                   'revenue',
                                                   'number_of_awards_won',
                                                   'number_of_nominations',
                                                   'number_of_top_productions']),
                                                 ('cat_transformer',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse=False),
                                                  ['original_language',
                                                   'status', 'has_collection',
                                                   'all_genres',
                                                   'top_countries',
                                                   'available_in_english'])])),
                ('linear_regression', Ridge())])

### Cross-validated baseline

**📝 Perform a cross-validated evaluation of your baseline model using the metric you defined above. Store the results of this evaluation as an `array` of floating scores in the `basic_scores` variable.**

In [90]:
# YOUR CODE HERE
from sklearn.model_selection import cross_val_score

# Cross validate pipeline
basic_scores = cross_val_score(final_pipe,X , y, cv=5, scoring='r2')

### 🧪 Save your results

Run the following cell to save your results

In [98]:
ChallengeResult(
    'baseline',
    metric=scoring,
    features=[categorical,numerical],
    preproc=basic_preprocessing,
    preproc_shape=X_basic_preprocessing.shape,
    pipe=basic_pipeline,
    scores=basic_scores
).write()

## Feature engineering

### Time Features


👉 Let's try to improve performance using the feature `release_date`, and especially its `month` and `year`.

ℹ️ If you want to skip this section, you can move directly to the next one: _Advanced categorical features_.

**📝 Complete the custom transformer `TimeFeaturesExtractor` below**

Running
```python
TimeFeaturesExtractor().fit_transform(X[['release_date']])
``` 
should return something like

|    |   month |   year |
|---:|--------:|-------:|
|  0 |       2 |   2015 |
|  1 |       8 |   2004 |
|  2 |      10 |   2014 |
|  3 |       3 |   2012 |
|  4 |       8 |   2012 |


In [110]:
daate = data['release_date'].iloc[0]
int(daate[:4])

2015

In [144]:
X=data['release_date']
def slicend(x):
    return x[5:7]
def slicebegin(x):
    return x[:4]

month = X.apply(sliced)


In [174]:
from sklearn.base import BaseEstimator, TransformerMixin

class TimeFeaturesExtractor(BaseEstimator, TransformerMixin):
    """Extract the 2 time features from a date"""
    
    def fit(self, X, y=None):
        return self
    
    def transform(self, X, y=None):
        """
        Params:
        X: DataFrame
        y: Series
        
        Returns a DataFrame with 2 columns containing the time features as integers extracted from the release_date.
        """
        arrm=[]
        arry=[]
        for i,d in X.iterrows():
            #print(d)
            
            year = d[:4]
            val = year[0]
            month = int(val[5:7])
            year = int(val[:4])
            arrm.append(month)
            arry.append(year)
        #print(month,year)
        d = {'month': arrm, 'year': arry}

        
        return pd.DataFrame(data=d)

In [175]:
# Try your transformer and save your new features here
X_time_features = TimeFeaturesExtractor().fit_transform(data[['release_date']])
X_time_features.head()

month  year
0      2  2015
1      8  2004
2     10  2014
3      3  2012
4      8  2012

We still have 2 problems to solve
- `month` is cyclical: 12 should be a close to 1 as to 11, right? 
- `year` is not scaled

**📝 Build a final custom transformer `CyclicalEncoder` so that**

Running
```python
CyclicalEncoder().fit_transform(X_time_features)
``` 
should return something like this

|    |    month_cos |   month_sin |      year |
|---:|-------------:|------------:|----------:|
|  0 |  0.5         |    0.866025 | 0.0466039 |
|  1 | -0.5         |   -0.866025 | 0.0411502 |
|  2 |  0.5         |   -0.866025 | 0.0461081 |
|  3 |  6.12323e-17 |    1        | 0.0451165 |
|  4 | -0.5         |   -0.866025 | 0.0451165 |

With the cyclical encoding is done as below
- `month_cos = 2 * math.pi / 12 * X[['month']] `
- `month_sin = 2 * math.pi / 12 * X[['month']] `

And the `year` begin min-max scaled

In [219]:
from sklearn.base import BaseEstimator, TransformerMixin
import math

class CyclicalEncoder(BaseEstimator, TransformerMixin):
    """
    Encode a cyclical feature
    """
    
    def __init__(self):
        pass

    def fit(self, X, y=None):
        """
        Compute here what you need for the transform phase and store it as instance variable
        """
        self.min = X['year'].min()
        self.max = X['year'].max()
        #self.year = (X – self.min) / (self.max – self.min)
        return self

    def transform(self, X, y=None):
        """
        Compute and returns the final DataFrame
        """
        
        #print(self.min)
        month_cos = 2 * math.pi / 12 * X[['month']]
        month_sin = 2 * math.pi / 12 * X[['month']]
        #print(month_cos,month_sin)
        year = (X[['year']] - self.min) / (self.max - self.min)
        #year = X[['year']]
        d = {'month_cos': month_cos,'month_sin' :month_sin, 'year': year}
        return  pd.DataFrame(data=d)


In [218]:
# Try your transformer and save your new features here
X_time_cyclical = CyclicalEncoder().fit_transform(X_time_features)
X_time_cyclical.head()

ValueError: If using all scalar values, you must pass an index

In [33]:
# Check that this form a circle with 12 points
plt.scatter(X_time_cyclical['month_cos'],
            X_time_cyclical['month_sin'])
plt.xlabel("month_cos"); plt.ylabel("month_sin");

**📝 Enhance your `basic_pipeline` with a new preprocessing including both `TimeFeaturesExtractor` and `CyclicalFeatureExtractor`:**

- Just use `TimeFeatureExtractor` if you haven't had time to do the `Cyclical` one
- Store this new pipeline as `time_pipeline`
- Keep same estimator for now

In [220]:
# YOUR CODE HERE
time_pipeline = Pipeline([
    
    ('TimeScaler', TimeFeaturesExtractor())])

### Advanced categorical encoder to reduce the number of features

ℹ️ Most of it has already been coded for you and it shouldn't take long. Still if you want to skip it and move to the next section: _Model Tuning_

👉 We need to reduce the number of features to one-hot-encode, which arise from the high cardinality of `all_genres` and `top_countries`

In [222]:
X = data

In [223]:
X[['all_genres', 'top_countries']].nunique()

all_genres       745
top_countries    320
dtype: int64

👇 Both share a common pattern: there can be more than 1 country and more than 1 genre per movie.

In [224]:
X[['all_genres', 'top_countries']].tail()

all_genres  \
6859             Animation, Drama, Family   
6860               Comedy, Drama, Romance   
6861                        Comedy, Drama   
6862  Adventure, Mystery, Science Fiction   
6863     Horror, Mystery, Science Fiction   

                                 top_countries  
6859                  United States of America  
6860                  United States of America  
6861                  United States of America  
6862  United Kingdom, United States of America  
6863  United Kingdom, United States of America

👉 Run the cell below where we have coded for you a custom transformer `CustomGenreAndCountryEncoder` which: 
- Select the 10 most frequent genres and the 5 most frequent countries
- Encode `all_genres` into 10 One Hot Encoded features
- Encode `top_countries` into 5 One Hot Encoded features

In [225]:
from collections import Counter
from sklearn.base import BaseEstimator, TransformerMixin

class CustomGenreAndCountryEncoder(BaseEstimator, TransformerMixin):
    """
    Encoding the all_genres and top_companies features which are multi-categorical :
    a movie has several possible genres and countries of productions!
    """

    def __init__(self):
        pass

    def fit(self, X, y=None):
        """
        compute top genres and top countries of productions from all_genres and top_countries features
        """

        # compute top 10 genres       
        list_of_genres = list(X['all_genres'].apply(lambda x: [i.strip() for i in x.split(",")] if x != [''] else []).values)
        top_genres = [m[0] for m in Counter([i for j in list_of_genres for i in j]).most_common(10)]

        # save top_genres in dedicated instance variable
        self.top_genres = top_genres
        
         # compute top 5 countries       
        list_of_countries = list(X['top_countries'].apply(lambda x: [i.strip() for i in x.split(",")] if x != [''] else []).values)
        top_countries = [m[0] for m in Counter([i for j in list_of_countries for i in j]).most_common(5)]

        # save top_countries in dedicated instance variable
        self.top_countries = top_countries

        return self

    def transform(self, X, y=None):
        """
        encoding genre and country
        """
        X_new = X.copy()
        for genre in self.top_genres:
            X_new['genre_' + genre] = X_new['all_genres'].apply(lambda x: 1 if genre in x else 0)
        X_new = X_new.drop(columns=["all_genres"])
        for country in self.top_countries:
            X_new['country_' + country] = X_new['top_countries'].apply(lambda x: 1 if country in x else 0)
        X_new = X_new.drop(columns=["top_countries"])
        return X_new

In [226]:
# Check it out
X_custom = CustomGenreAndCountryEncoder().fit_transform(X[['all_genres', 'top_countries']])
print(X_custom.shape)
X_custom.head()

(6864, 15)


genre_Drama  genre_Comedy  genre_Thriller  genre_Action  genre_Romance  \
0            0             1               0             0              0   
1            1             1               0             0              1   
2            1             0               0             0              0   
3            1             0               1             0              0   
4            0             0               1             0              0   

   genre_Adventure  genre_Crime  genre_Horror  genre_Science Fiction  \
0                0            0             0                      0   
1                0            0             0                      0   
2                0            0             0                      0   
3                0            0             0                      0   
4                0            0             1                      0   

   genre_Family  country_United States of America  country_United Kingdom  \
0             0                                 1                       0   
1             1                                 1                       0   
2             0                                 1                       0   
3             0                                 0                       0   
4             0                                 1                       0   

   country_France  country_Germany  country_Canada  
0               0                0               0  
1               0                0               0  
2               0                0               0  
3               0                0               0  
4               0                0               1

**📝 Compute your `final_pipeline` by integrating all these transformers** (or all those you have coded)

- `CustomGenreAndCountryEncoder`
- `TimeFeaturesExtractor`
- `CyclicalFeatureExtractor`

In [231]:
# YOUR CODE HERE
categoricales = ['original_language','status','has_collection','available_in_english']
preprocessor = ColumnTransformer([
    ('num_transformer', num_transformer, numerical),
    ('TimeScaler', TimeFeaturesExtractor(),dates),
    ('custom_encode',CustomGenreAndCountryEncoder(),['all_genres', 'top_countries']),
    ('cat_transformer', cat_transformer, categoricales)]
)
final_pipe = Pipeline([
    ('preprocessing', preprocessor),
    
    ('linear_regression', Ridge())])
final_pipe

Pipeline(steps=[('preprocessing',
                 ColumnTransformer(transformers=[('num_transformer',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['duration_min', 'budget',
                                                   'revenue',
                                                   'number_of_awards_won',
                                                   'number_of_nominations',
                                                   'number_of_top_productions']),
                                                 ('TimeScaler',
                                                  TimeFeaturesExtractor(),
                                                  ['release_date']),
                                                 ('custom_encode',
                                                  CustomGenreAndCountryEncoder(),
                                                  ['all_genres',
                                                   'top_countries']),
                                                 ('cat_transformer',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse=False),
                                                  ['original_language',
                                                   'status', 'has_collection',
                                                   'available_in_english'])])),
                ('linear_regression', Ridge())])

In [232]:
final_scores = cross_val_score(final_pipe,X , y, cv=5, scoring='r2')

In [234]:
final_scores

array([0.18959227, 0.17660596, 0.22923365, 0.1693959 , 0.07797017])

📝 **Compute and store its cross validated scores as `final_scores` array of floats**

- It does not necessarily improve the performance before we can try-out doing model tuning
- However, with a now limited number of features, we will be able to train more complex models in next section (ensemble...)

### 🧪 Save your result

Run the following cell to save your results.

In [235]:
ChallengeResult(
    'feature_engineering',
    X_time_features=X_time_features,
    X_time_cyclical= X_time_cyclical,
    time_pipeline=time_pipeline,
    final_pipeline=final_pipeline,
    final_scores=final_scores
).write()

# Hint: Try restarting your notebook if you obtain an error about saving a custom encoder

NameError: name 'X_time_cyclical' is not defined

## Model tuning

### Random Forest

📝 **Change the estimator of your `final_pipeline` by a Random Forest and checkout your new cross-validated score**

In [237]:
# YOUR CODE HERE
from sklearn.ensemble import RandomForestRegressor
final_pipeline = Pipeline([
    ('preprocessing', preprocessor),
    
    ('raandom_forest', RandomForestRegressor())])
final_pipeline
cross_val_score(final_pipeline,X , y, cv=5, scoring='r2')

array([0.47309495, 0.33061254, 0.25677895, 0.16772204, 0.25863439])

### Best hyperparameters quest



**📝 Fine tune your model to try to get the best performance in the minimum amount of time!**

- Store the result of your search inside the `search` variable.
- Store your 5 cross-validated scores inside `best_scores` array of floats

In [47]:
# YOUR CODE HERE
from sklearn.model_selection import GridSearchCV

# Instanciate grid search
grid_search = GridSearchCV(
    final_pipeline, 
    param_grid={
        # Access any component of the pipeline, as far back as you want
        'preprocessing__num_transformer__imputer__strategy': ['mean', 'median'],
        'preprocessing__children_scaler__shrink_factor': [1,2,3,4],
        'linear_regression__alpha': [0.1, 0.5, 1, 5, 10]},
    cv=5,
    scoring="r2")

grid_search.fit(X_train, y_train)
grid_search.best_params_

**📝 Re-train your best pipeline on the whole (X,y) dataset**
- Store the trained pipeline inside the `best_pipeline` variable

In [53]:
# YOUR CODE HERE

### Prediction

Now you have your model tuned with the best hyperparameters, you are ready for a prediction.

Here is a famous TV show released in 2017:

```python
dict(
        original_title=str("La Casa de Papel"),
        title=str("Money Heist"), 
        release_date= pd.to_datetime(["2017-05-02"]), 
        duration_min=float(50),
        description=str("An unusual group of robbers attempt to carry out the most perfect robbery"), 
        budget=float(3_000_000), 
        original_language =str("es"), 
        status=str("Released"),
        number_of_awards_won =int(2), 
        number_of_nominations=int(5), 
        has_collection=int(1),
        all_genres=str("Action, Crime, Mystery"), 
        top_countries=str("Spain, France, United States of America"), 
        number_of_top_productions=int('1'),
        available_in_english=bool('True') 
)
```

**📝 Compute the predicted popularity of this TV show and store it into the `popularity` variable as a floating number.**

In [54]:
# YOUR CODE HERE

In [55]:
# YOUR CODE HERE

### 🧪 Save your results

Run the following cell to save your results.

In [56]:
ChallengeResult(
    "model_tuning",
    search=search,
    best_pipeline=best_pipeline,
    best_scores = best_scores,
    popularity=popularity
).write()

## API 

Time to put a pipeline in production!

👉 Go to https://github.com/lewagon/data-certification-api and follow instructions

**This final part is independent from the above notebook**